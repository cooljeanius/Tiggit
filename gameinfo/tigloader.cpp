#include "tigloader.hpp"

#include <misc/readjson.hpp>
#include <boost/filesystem.hpp>
#include <stdexcept>
#include <cstdlib>

using namespace GameInfo;
using namespace TigData;
using namespace boost::filesystem;

static void fail(const std::string &msg)
{
  throw std::runtime_error(msg);
}

void TigLoader::clear()
{
  lookup.clear();
  channels.clear();
}

static std::string getNext(const char* &ptr)
{
  while(*ptr == ' ') ptr++;
  const char *start = ptr;
  while(*ptr != ' ' && *ptr != 0) ptr++;
  return std::string(start, ptr-start);
}

void TigLoader::addChannel(const std::string &channel,
                           const std::string &jsonfile)
{
  // Check the channel name
  if(channel.find('\\') != std::string::npos ||
     channel.find('/') != std::string::npos)
    fail("Invalid channel name '" + channel + "' (empty or contains slashes)");

  // Does the channel already exist?
  if(getChannel(channel) != NULL)
    fail("Channel '" + channel + "' already exists!");

  /* TODO: In future versions, we may add some structure around the
     list, maybe include the channel name and other info in the file
     itself. It's important that we are able to catch and throw on
     version updates, because the client relies on exceptions to
     signal that the data is in an outdated format.
   */

  // Load the JSON data
  Json::Value root = ReadJson::readJson(jsonfile);
  if(root.isNull()) return;
  if(!root.isArray())
    fail("Invalid game data in " + jsonfile);

  // Create and set up the list
  ListPtr list(new TigList);
  list->channel = channel;

  list->list.reserve(root.size());
  for(int i=0; i<root.size(); i++)
    {
      const Json::Value &v = root[i];
      TigEntry e;

      e.launch = v["launch"].asString();
      e.title = v["title"].asString();
      e.desc = v["desc"].asString();
      e.devname = v["devname"].asString();
      e.homepage = v["homepage"].asString();
      e.tags = v["tags"].asString();
      e.urlname = v["name"].asString();

      {
        std::string libs = v["libs"].asString();
        const char *ptr = libs.c_str();
        for(;;)
          {
            std::string next = getNext(ptr);
            if(next == "") break;
            e.libs.push_back(next);
          }
      }

      e.channel = channel;
      e.idname = channel + "/" + e.urlname;
      e.flags = 0;

      if(v["is_demo"].asBool())
        e.flags |= TF_DEMO;

      e.addTime = std::atoll(v["addtime"].asString().c_str());

      /* TODO: Perform more sanitation here. The tigdata is in
         principle untrusted data.
       */
      if(e.launch == "" || e.title == "" || e.urlname == "")
        continue;

      // Add the item if it checked out
      list->list.push_back(e);
    }

  // Store the list, and add all the entries to the lookup
  channels[channel] = list;
  for(int i=0; i<list->list.size(); i++)
    {
      TigEntry *e = &list->list[i];
      lookup[e->idname] = e;
    }
}

const TigEntry* TigLoader::getGame(const std::string &idname) const
{
  Lookup::const_iterator it = lookup.find(idname);
  if(it == lookup.end())
    return NULL;
  return it->second;
}

TigEntry* TigLoader::editGame(const std::string &idname)
{
  Lookup::iterator it = lookup.find(idname);
  if(it == lookup.end())
    return NULL;
  return it->second;
}

const TigList* TigLoader::getChannel(const std::string &name) const
{
  ChanMap::const_iterator it = channels.find(name);
  if(it == channels.end())
    return NULL;
  return it->second.get();
}
