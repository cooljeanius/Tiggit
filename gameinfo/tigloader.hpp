#ifndef __GI_TIGLOADER_HPP_
#define __GI_TIGLOADER_HPP_

#include "tiglist.hpp"
#include <map>
#include <boost/shared_ptr.hpp>

namespace GameInfo
{
  class TigLoader
  {
  public:
    typedef boost::shared_ptr<TigData::TigList> ListPtr;
    typedef std::map<std::string, TigData::TigEntry*> Lookup;
    typedef std::map<std::string, ListPtr> ChanMap;

  private:

    Lookup lookup;
    ChanMap channels;

  public:
    /* Load a channel data file. An exception is thrown if the name
       matches an already existing channel name, or if any other error
       occurs.
     */
    void addChannel(const std::string &channel,
                    const std::string &dataFile);

    // Clear away all loaded data
    void clear();

    /* Find a game based on idname ("channel/name"). Returns NULL if
       not found.
     */
    const TigData::TigEntry* getGame(const std::string &idname) const;

    /* Edit a game entry. Used by the stats loader. */
    TigData::TigEntry *editGame(const std::string &idname);

    /* Find a channel's info. Returns NULL if not found.
     */
    const TigData::TigList* getChannel(const std::string &name) const;

    /* Return the full lookup of games
     */
    const Lookup &getList() const { return lookup; }

    /* Return full channel list
     */
    const ChanMap &getChanList() const { return channels; }
  };
}

#endif
