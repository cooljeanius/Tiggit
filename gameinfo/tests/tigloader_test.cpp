#include "tigloader.hpp"
#include <iostream>

using namespace GameInfo;
using namespace std;

int main()
{
  TigLoader tig;

  tig.addChannel("data.bin");

  const TigLoader::Lookup &list = tig.getList();
  TigLoader::Lookup::const_iterator it;
  for(it = list.begin(); it != list.end(); it++)
    {
      const TigData::TigEntry &ent = *it->second;
      cout << "\n" << ent.urlname << ":\n";
      cout << "  Title: " << ent.title
           << "\n  Homepage: " << ent.homepage
           << "\n  Tags: " << ent.tags
           << "\n  Channel: " << ent.channel
           << "\n  Rating: " << ent.rating << " (" << ent.rateCount << " votes)\n";
      cout << "  Demo: " << (ent.isDemo()?"yes":"no") << endl;
    }

  return 0;
}
