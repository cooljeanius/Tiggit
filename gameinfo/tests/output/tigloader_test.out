
brainsplode:
  Title: BrainSplode!
  Homepage: http://www.richmakegame.com
  Tags: single-player shooter turn-based puzzle physics
  Channel: test_channel
  Rating: -1 (0 votes)
  Demo: no

brogue:
  Title: Brogue
  Homepage: https://sites.google.com/site/broguegame/home
  Tags: single-player roguelike role-playing adventure
  Channel: test_channel
  Rating: -1 (0 votes)
  Demo: no

moon-invaders:
  Title: Moon Invaders
  Homepage: http://hamumu.com/game.php?game=MOON
  Tags: single-player space-invaders strategy retro
  Channel: test_channel
  Rating: -1 (0 votes)
  Demo: no

aquaria---demo:
  Title: Aquaria - DEMO
  Homepage: http://www.bit-blot.com/aquaria/buy.html
  Tags: adventure action single-player
  Channel: tiggit.net
  Rating: 4.37037 (27 votes)
  Demo: yes

cave-story:
  Title: Cave Story
  Homepage: http://www.cavestory.org/
  Tags: single-player platform adventure action
  Channel: tiggit.net
  Rating: 4.62585 (394 votes)
  Demo: no

dwarf-fortress:
  Title: Dwarf Fortress
  Homepage: http://www.bay12games.com/dwarves/
  Tags: strategy ascii procedural fantasy sandbox real-time single-player
  Channel: tiggit.net
  Rating: 4.52705 (227 votes)
  Demo: no
